cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java

```diff
@@ -601,7 +570,11 @@ public Void call() throws Exception {
               }
             }
             context.setState(getProgramState(success, failureInfo));
-            ((ProgramLifecycle) mapReduce).destroy();
+            if (mapReduce instanceof ProgramLifecycle) {
```

Comment:
I think we need to improve the exception handling here. But that can be done as part of CDAP-4065. So it's fine for now.